### PR TITLE
Recipient dashboard

### PR DESCRIPTION
**Pull request by qdozaq**
submitted
Annotation 2020-05-21 185605

### CODE REVIEW

File: src/app/page/Recipient/Dashboard.tsx [hunk 1]
```diff
@@ -28,31 +28,41 @@ const useStyles = makeStyles((theme) => ({
...
+      let submitted: MissionInterface[] = [];
+      let completed: MissionInterface[] = [];
 
       missions.forEach((mission) => {
         completedStatus.includes(mission.status)
           ? completed.push(mission)
           : submitted.push(mission);
       });
 
+      submitted.sort(sortByDate);
```
Comment (mat10tng) [inline src/app/page/Recipient/Dashboard.tsx:63]: you could sort the missions first, then both submitted and completed would be sort by default when they are pushed in

File: src/app/page/Recipient/Details.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    <>
+      <Divider />
+      <Box margin="1rem">
+        {mission.deliveryType === "delivery" ? (
+          <>
+            <DetailSection
+              header="Address"
+              content={
+                <Typography variant="body1" gutterBottom>
+                  {mission.deliveryLocation.address}
```
Comment (mat10tng) [inline src/app/page/Recipient/Details.tsx:21]: mission?.

Comment (qdozaq) [reply src/app/page/Recipient/Details.tsx:21]: I can, but typescript guarantees that mission will exist here

File: src/app/model/Mission.ts [hunk 3]
```diff
@@ -265,7 +266,7 @@ class Mission extends BaseModel {
    * @param {string} missionUid - mission
    * @param {object} data- updated data
    */
-  update(missionUid: string, data: object) {
+  update(missionUid: string, data: Partial<MissionInterface>) {
```
Comment (qdozaq) [inline src/app/model/Mission.ts:269]: Special TS wrapper to make all properties optional

File: src/app/routing/components/AppRoute.tsx [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
 const loadingStyles = makeStyles((theme) => ({
   root: {
     width: "100%",
   },
 }));
 
-function AppRoute({ children, component, ...rest }) {
+type Props = RouteProps;
+
+function AppRoute({ children, component, ...rest }: Props) {
```
Comment (qdozaq) [inline src/app/routing/components/AppRoute.tsx:19]: @amylo I converted this component to typescript. lemme know if I'm missing anything

File: src/app/model/schema.ts [hunk 5]
```diff
@@ -215,7 +216,7 @@ export interface MissionInterface {
   deliveryLocation: Location; // default to recipient location
   deliveryConfirmationImage: string;
   deliveryNotes: string;
-  deliveryType: string;
+  deliveryType: "delivery" | "curbside";
```
Comment (jwu910) [inline src/app/model/schema.ts:219]: :eyes:

File: src/app/page/Recipient/Details.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        )}
+      </Box>
+    </>
+  );
+}
+
+function DetailSection({ content, header }: { content: ReactNode; header: ReactNode }) {
+  return (
+    <Box marginBottom="1rem">
+      {React.isValidElement(header) ? header : <Typography variant="h5">{header}</Typography>}
```
Comment (jwu910) [inline src/app/page/Recipient/Details.tsx:65]: Oh interesting validation to check for jsx vs just a text string?

File: src/app/page/Recipient/Details.tsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    if (!isEditing) {
+      setIsEditing(true);
+      return;
+    }
+
+    Mission.update(mission.uid, { deliveryNotes })
+      .then(() => {
+        setIsEditing(false);
+      })
+      .catch((e) => {
```
Comment (jwu910) [inline src/app/page/Recipient/Details.tsx:87]: Do we want to provide `e.message` for any reason to the user or organizer? Is it possible Mission.update would have any valuable info that they could resolve?

File: src/app/model/schema.ts (hunk 5 above)
Comment (qdozaq) [reply src/app/model/schema.ts:219]: its like a poor mans enum. Really nice cause it auto fills these options when either filling the prop or using it to compare

File: src/app/page/Recipient/Details.tsx (hunk 6 above)
Comment (qdozaq) [reply src/app/page/Recipient/Details.tsx:65]: yah!

File: src/app/page/Recipient/RequestsList.tsx [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+  const [open, setOpen] = useState(false);
+  const toggle = () => setOpen(!open);
+
+  return (
+    <Card className={classes.card}>
+      <TopBar
+        status={mission.status}
+        fundedStatus={mission.fundedStatus}
+        createdDate={mission.createdDate}
+      ></TopBar>
```
Comment (jwu910) [inline src/app/page/Recipient/RequestsList.tsx:109]: Just syntax, but can we just self close if there are no children?

File: src/app/page/Recipient/RequestsList.tsx [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+  ) : status === Status.delivered ? (
+    <NotificationImportant {...rest} />
+  ) : status === Status.succeeded ? (
+    <CheckCircle {...rest} />
+  ) : fundedStatus === FundedStatus.notfunded ? (
+    <HourglassFull {...rest} />
+  ) : (
+    <FiberNew {...rest} />
+  );
+}
```
Comment (jwu910) [inline src/app/page/Recipient/RequestsList.tsx:154]: My eyes! haha jk

Can we just use a hash map(?) like this:

```
const ICON_STATUS = {
  [Status.started]: PlayCircleFilled,
  [Status.delivered]: <NotificationImportant,
  [Status.succeeded]: CheckCircle,
  [FundedStatus.notfunded]: HourglassFull
  ['default']: FiberNew
}

// and then use it something like
const StatusIcon = (status) => {
	if (!(Object.keys(ICON_STATUS).contains(status)) {
		return ICON_STATUS['default']
	}

    return ICON_STATUS[status]
}

//and then 
//...

<StatusIcon [...your props ]/>
```

Some of these ternary chains are really hard to follow

File: src/app/page/Recipient/Details.tsx (hunk 7 above)
Comment (qdozaq) [reply src/app/page/Recipient/Details.tsx:87]: I wish we had a logging service... I can `console.error` it but I don't think customers need to see the raw message

File: src/app/page/Recipient/RequestsList.tsx (hunk 9 above)
Comment (qdozaq) [reply src/app/page/Recipient/RequestsList.tsx:154]: I wanted to do a map or even just a switch but I need to account for two different status objects which causes problems... like when would fundedstatus override the regular status?

Would and `if/else` look cleaner?

```ts
if (status === STATUS...) return <Icon/>
if (status === STATUS...) return <Icon/>
if (status === STATUS...) return <Icon/>
if (status === STATUS...) return <Icon/>
if (fundedstatus === FUNDEDSTATUS...) return <Icon/>
```

File: src/app/model/schema.ts (hunk 5 above)
Comment (amylo) [reply src/app/model/schema.ts:219]: Might as well actually define String Enum lol.... Cuz then we can share and reuse elsewhere when needed. Don't want to give people reason to cooy-pasta.

File: src/app/page/Recipient/Details.tsx (hunk 2 above)
Comment (amylo) [reply src/app/page/Recipient/Details.tsx:21]: We probably have bigger issues, if a required property is missing 😉 should trust TypeScript to enforce the requirement.

File: src/app/model/schema.ts (hunk 5 above)
Comment (qdozaq) [reply src/app/model/schema.ts:219]: Haha okay. Fair enough

File: src/app/page/Recipient/RequestsList.tsx (hunk 9 above)
Comment (mat10tng) [reply src/app/page/Recipient/RequestsList.tsx:154]: I would prefer switch due to both:
1. Switch is made for this use case, map is just a convenient way.
2. You actually create an object every time you render using the map approach, we had this discussion earlier.

Comment (qdozaq) [reply src/app/page/Recipient/RequestsList.tsx:154]: neither the switch nor map will work unless we can consolidate all statuses to single property.
Need to be able to test for both status and funded status which I can't really directly do in a switch.
Gonna try the `if` block approach

File: src/app/model/schema.ts (hunk 5 above)
Comment (jwu910) [reply src/app/model/schema.ts:219]: I threw the eyes in because this was changed from `"delivery" | "curbside"` to `string` a few days ago and now back to this lol

Comment (qdozaq) [reply src/app/model/schema.ts:219]: Oh lol...
It should definitely be a limited set of values as we're using it to determine UI state.

File: src/app/model/Mission.ts [hunk 10]
```diff
@@ -360,6 +362,7 @@ class Mission extends BaseModel {
       volunteerUid: userUid,
       volunteerDisplayName: user.displayName,
       volunteerPhoneNumber: user.phoneNumber,
+      volunteerPhotoUrl: user.photoURL,
```
Comment (mat10tng) [inline src/app/model/Mission.ts:365]: volunteerPhotoURL is more consistent

Comment (qdozaq) [reply src/app/model/Mission.ts]: done

File: src/app/page/Recipient/RequestsList.tsx [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+function TopBarText({ createdDate, fundedStatus, status }: any) {
+  const date = new Date(createdDate).toLocaleDateString();
+
+  if (status === Status.started) return <>In progress | Accepted</>;
+  if (status === Status.delivered) return <>It's here! Confirm delivery below</>;
+  if (status === Status.succeeded) return <>Delivery Confirmed</>;
+  if (fundedStatus === FundedStatus.notfunded) return <>Awaiting donation | Submitted {date}</>;
+  return <>Submitted {date}</>;
```
Comment (jwu910) [inline src/app/page/Recipient/RequestsList.tsx:157]: Lol Hmm i mean it... works.. It's actually much easier to read with this many if's. I just feel like this can be a littler DRY-er somehow... But i think we're good for now. 

Does this need to return text in a fragment though? Should that just be a string?

Comment (qdozaq) [reply src/app/page/Recipient/RequestsList.tsx]: I was tempted to share the Icon and text state in the same component as they share the same return states, but that isn't necessarily guaranteed going forward as the designs don't really show all the possible combinations that the top bar can be in. Keeping it separate should make future changes easier if/when these states diverge.

I had initially used fragment as typescript has an issue since `string` isn't a jsx constructor. Apparently still an ongoing issue (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20544)

Went back and was able to cheat it with `any` though so guess that works lol

**Comment by jwu910** [comment]
Ticket #?
